crypto/merklearray/merkle.go

```diff
@@ -0,0 +1,133 @@
+package merklearray
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+type Tree struct {
+	// Level 0 is the leaves.
+	levels []layer
+}
+
+func (tree *Tree) topLayer() layer {
+	return tree.levels[len(tree.levels)-1]
+}
+
+func Build(array Array) (*Tree, error) {
+	tree := &Tree{}
+
+	var leaves layer
+	arraylen := array.Length()
+	for i := uint64(0); i < arraylen; i++ {
+		data, err := array.Get(i)
+		if err != nil {
+			return nil, err
+		}
+
+		leaves = append(leaves, crypto.HashObj(data))
+	}
+
+	if arraylen > 0 {
+		tree.levels = []layer{leaves}
```

Comment:
you can pretty easily preallocate the levels array.

Reply:
I think it would need to be something like `1+floor(log2(len(leaves)-1)`, which seems messy enough, combined with the lack of `log2()` for integers in Go, to not bother with the preallocation..